src/LevelZeroDevicePoolImp.hpp

```diff
@@ -0,0 +1,143 @@
+            virtual double standby_mode(unsigned int accel_idx) const override;
+            virtual double memory_allocated(unsigned int accel_idx) const override;
+            virtual void energy_threshold_control(unsigned int accel_idx, double setting) const override;
+            virtual void frequency_control_gpu(unsigned int accel_idx,  double setting) const override;
+            virtual void standby_mode_control(unsigned int accel_idx, double setting) const override;
+
+        private:
+            const unsigned int M_NUM_CPU;
+            virtual void domain_cache(unsigned int accel_idx);
+            virtual void check_accel_range(unsigned int accel_idx) const;
+            virtual void check_domain_range(int size, const char *func, int line) const;
+            virtual void check_ze_result(ze_result_t ze_result, int error, std::string message, int line) const;
+            virtual int num_accelerator(ze_device_type_t type) const;
+
+            virtual std::tuple<zes_power_sustained_limit_t,
+                               zes_power_burst_limit_t,
+                               zes_power_peak_limit_t> power_limit(unsigned int accel_idx) const;
+
+            virtual std::tuple<int32_t, int32_t, int32_t> power_limit_default(unsigned int accel_idx) const;
+            virtual double frequency_status(int accel_idx, zes_freq_domain_t) const;
+            virtual std::pair<double, double> frequency_min_max(int accel_idx, zes_freq_domain_t type) const;
+            virtual std::pair<double, double> frequency_range(int accel_idx, zes_freq_domain_t type) const;
+            virtual std::pair<uint64_t, uint64_t> energy_pair(unsigned int accel_idx) const;
+            virtual std::pair<uint64_t, uint64_t> active_time(unsigned int accel_idx, zes_engine_group_t engine_type) const;
+            virtual double temperature(int accel_idx, zes_temp_sensors_t sensor_type) const;
+            virtual void frequency_control(unsigned int accel_idx, double min_freq, double max_freq, zes_freq_domain_t) const;
+
+            std::string ze_error_string(ze_result_t result);
+
+            uint32_t m_num_driver;
+            uint32_t m_num_device;
+            uint32_t m_num_integrated_gpu;
+            uint32_t m_num_board_gpu;
+            uint32_t m_num_fpga;
+            uint32_t m_num_mca;
+
+            std::vector<ze_driver_handle_t> m_levelzero_driver;
+            // TODO: This will need to be a map or vector of vectors if we're supporting multiple
+            //       types of sysman devices such as <BOARD GPU devices, PACKAGE GPU device, FPGA device>
+            std::vector<zes_device_handle_t> m_sysman_device;
+
+            // TODO: These will need to be something like a map of vector of vectors if we're supporting multiple
+            //       types of sysman devices such as <Temperature<device<Board GPU, Package GPU, FPGA>>>
+            std::vector<std::vector<zes_freq_handle_t> > m_freq_domain;
```

Comment:
Consider moving tracking to a device level struct, such as:
`device_struct{
    device_handles;
    domain_handles[power,freq,engine,...]
    device_type;
    `